File: src/features/games/engine/engine.ts
/**
 * Upmind — game engine core.
 * Manages trial state, RT, drift detection, and the trial lifecycle.
 * One session = one game + N trials. The GamePlayer screen is a thin
 * UI shell that calls into the engine.
 */

import { GameDef, getGame } from '../catalog';
import { Trial } from './types';
import { generators } from './generators';

export type AnswerRecord = {
  trialIndex: number;
  rtMs: number;
  correct: boolean;
  /** User's raw response (for reanalysis). */
  response: unknown;
  /** True if this trial was flagged as anti-cheat drift. */
  drift: boolean;
};

export type SessionResult = {
  gameId: string;
  construct: string;
  startedAt: number;
  finishedAt: number;
  trials: Trial[];
  answers: AnswerRecord[];
  score: number;
  rtMedianMs: number;
  rtStddevMs: number;
  accuracy: number;
  drifts: number;
};

export type EngineState = {
  game: GameDef;
  difficulty: number;
  trials: Trial[];
  answers: AnswerRecord[];
  currentIndex: number;
  startTime: number;
  trialStart: number;
  isFinished: boolean;
  isStarted: boolean;
  drifts: number;
};

export type EngineListeners = {
  onTrial: (trial: Trial, index: number) => void;
  onAnswer: (record: AnswerRecord) => void;
  onFinish: (result: SessionResult) => void;
};

const RT_IMPOSSIBLE_FLOOR_MS = 100;

export class Engine {
  state: EngineState;
  listeners: Partial<EngineListeners>;

  constructor(game: GameDef, listeners: Partial<EngineListeners> = {}) {
    this.state = {
      game,
      difficulty: 1,
      trials: [],
      answers: [],
      currentIndex: 0,
      startTime: 0,
      trialStart: 0,
      isFinished: false,
      isStarted: false,
      drifts: 0,
    };
    this.listeners = listeners;
  }

  start() {
    const generator = generators[this.state.game.id];
    if (!generator) throw new Error(`No generator for game: ${this.state.game.id}`);
    const trials: Trial[] = [];
    for (let i = 0; i < this.state.game.trials; i++) {
      trials.push(generator(i, this.state.difficulty));
    }
    this.state.trials = trials;
    this.state.isStarted = true;
    this.state.startTime = Date.now();
    this.nextTrial();
  }

  private nextTrial() {
    if (this.state.currentIndex >= this.state.trials.length) {
      this.finish();
      return;
    }
    const trial = this.state.trials[this.state.currentIndex];
    this.state.trialStart = Date.now();
    this.listeners.onTrial?.(trial, this.state.currentIndex);
  }

  /** Returns true if correct. */
  answer(response: unknown): { correct: boolean; drift: boolean } {
    if (this.state.isFinished || !this.state.isStarted) {
      return { correct: false, drift: false };
    }
    const trial = this.state.trials[this.state.currentIndex];
    const rtMs = Date.now() - this.state.trialStart;
    const correct = this.isCorrect(trial, response);
    const drift = this.detectDrift(rtMs, response);

    if (drift) this.state.drifts++;

    const record: AnswerRecord = {
      trialIndex: this.state.currentIndex,
      rtMs,
      correct,
      response,
      drift,
    };
    this.state.answers.push(record);
    this.listeners.onAnswer?.(record);

    this.state.currentIndex++;
    this.nextTrial();
    return { correct, drift };
  }

  /** User exited the game before completing it. We discard partial data. */
  abort() {
    this.state.isFinished = true;
  }

  private isCorrect(trial: Trial, response: unknown): boolean {
    switch (trial.template) {
      case 'CHOICE':
      case 'RECALL': {
        const t = trial;
        const choice = t.choices.find((c) => c.id === response);
        return !!choice?.correct;
      }
      case 'REACTION': {
        const t = trial;
        return Boolean(response) === t.shouldPress;
      }
      case 'SEQUENCE': {
        const t = trial;
        const user = Array.isArray(response) ? (response as string[]) : [];
        if (user.length !== t.answer.length) return false;
        return user.every((v, i) => v === t.answer[i]);
      }
      case 'GRID': {
        const t = trial;
        if (!response || typeof response !== 'object') return false;
        const r = response as { r: number; c: number };
        return r.r === t.answer.r && r.c === t.answer.c;
      }
      case 'NUMBERLINE': {
        const t = trial;
        const user = typeof response === 'number' ? response : 0;
        return Math.abs(user - t.target) <= (t.max - t.min) * 0.05;
      }
      case 'TYPE': {
        const t = trial;
        const user = typeof response === 'string' ? response.trim().toLowerCase() : '';
        try {
          return new RegExp(t.answerPattern, 'i').test(user);
        } catch {
          return false;
        }
      }
      case 'SORT': {
        const t = trial;
        return response === t.answerIndex;
      }
    }
  }

  private detectDrift(rtMs: number, response: unknown): boolean {
    // RT below impossible-human floor
    if (rtMs < RT_IMPOSSIBLE_FLOOR_MS) return true;
    // Repeated exact RT (suspicious)
    const last = this.state.answers[this.state.answers.length - 1];
    if (last && last.rtMs === rtMs && JSON.stringify(last.response) === JSON.stringify(response)) {
      return true;
    }
    return false;
  }

  private finish() {
    this.state.isFinished = true;
    const finishedAt = Date.now();
    const correct = this.state.answers.filter((a) => a.correct).length;
    const score = Math.round((correct / this.state.answers.length) * 100);
    const rts = this.state.answers.map((a) => a.rtMs).sort((a, b) => a - b);
    const median = rts.length ? rts[Math.floor(rts.length / 2)] : 0;
    const mean = rts.length ? rts.reduce((a, b) => a + b, 0) / rts.length : 0;
    const variance = rts.length
      ? rts.reduce((acc, v) => acc + (v - mean) ** 2, 0) / rts.length
      : 0;
    const stddev = Math.sqrt(variance);

    const result: SessionResult = {
      gameId: this.state.game.id,
      construct: this.state.game.construct,
      startedAt: this.state.startTime,
      finishedAt,
      trials: this.state.trials,
      answers: this.state.answers,
      score,
      rtMedianMs: median,
      rtStddevMs: Math.round(stddev),
      accuracy: correct / Math.max(1, this.state.answers.length),
      drifts: this.state.drifts,
    };
    this.listeners.onFinish?.(result);
  }
}

export function createEngine(gameId: string, listeners?: Partial<EngineListeners>): Engine {
  const game = getGame(gameId);
  if (!game) throw new Error(`Unknown game: ${gameId}`);
  return new Engine(game, listeners);
}

File: ios/Upmind/Engine/Engine.swift
import Foundation

/// Drives one cognitive-training session. The engine is framework-agnostic
/// (no SwiftUI / UIKit imports) so it can be tested in isolation and
/// potentially reused on a watchOS app later.
actor Engine {

    enum Event: Sendable {
        case trial(Trial, index: Int)
        case answer(AnswerRecord)
        case finish(SessionResult)
    }

    private(set) var state: EngineState

    /// Stream of events emitted during the session. Consumers iterate with
    /// `for await event in engine.events`. Yielded non-blockingly from inside
    /// the actor; consumers never need to `await` an emit.
    private let eventStream: AsyncStream<Event>
    private let eventContinuation: AsyncStream<Event>.Continuation

    /// Drift floor: reaction times below 180ms are flagged as impossible-human.
    /// (Humans don't go faster than ~200ms for simple visual RT.)
    private let rtImpossibleFloorMs: Int = 180
    /// Drift streak cap: walk back at most this many prior answers when looking
    /// for a streak of identical RT + response.
    private let driftStreakCap: Int = 5
    /// Aggregate flag: a session whose median RT sits below this threshold is
    /// impossibly fast, so we add one extra drift to the count.
    private let aggregateMedianFloorMs: Int = 250

    init(
        game: GameDef,
        userIdentifier: String = "anonymous"
    ) throws {
        guard Games.game(game.id) != nil else {
            throw EngineError.unknownGame(game.id.rawValue)
        }
        self.state = EngineState(game: game, userIdentifier: userIdentifier)
        let (stream, continuation) = AsyncStream<Event>.makeStream()
        self.eventStream = stream
        self.eventContinuation = continuation
    }

    /// Public stream of events. Iterate from the UI / ViewModel with
    /// `for await event in engine.events { ... }`.
    nonisolated var events: AsyncStream<Event> { eventStream }

    // MARK: - Lifecycle

    func start() async {
        guard !state.isStarted, !state.isFinished else { return }
        guard let generator = Generators.lookup(state.game.id) else {
            state.isFinished = true
            return
        }
        var trials: [Trial] = []
        trials.reserveCapacity(state.game.trials)
        for i in 0..<state.game.trials {
            trials.append(generator.makeTrial(index: i, difficulty: state.difficulty))
        }
        state.trials = trials
        state.isStarted = true
        state.startTime = Date()
        state.trialStart = ContinuousClock.now
        emitNextTrial()
    }

    func answer(_ response: TrialResponse) async throws {
        guard state.isStarted else { throw EngineError.noActiveTrial }
        guard !state.isFinished else { throw EngineError.sessionAlreadyFinished }
        guard state.currentIndex < state.trials.count else {
            throw EngineError.duplicateAnswer
        }
        let trial = state.trials[state.currentIndex]
        let now = ContinuousClock.now
        let elapsed = now - state.trialStart
        let rtMs = Self.durationToMs(elapsed)
        let correct = try isCorrect(trial: trial, response: response)
        let drift = detectDrift(rtMs: rtMs, response: response)
        if drift { state.drifts += 1 }
        let record = AnswerRecord(
            trialIndex: state.currentIndex,
            rtMs: rtMs,
            correct: correct,
            response: response,
            drift: drift
        )
        state.answers.append(record)
        eventContinuation.yield(.answer(record))
        state.currentIndex += 1
        if state.currentIndex >= state.trials.count {
            finish()
        } else {
            emitNextTrial()
        }
    }

    func abort() {
        state.isFinished = true
        eventContinuation.finish()
    }

    // MARK: - Private

    private func emitNextTrial() {
        state.trialStart = ContinuousClock.now
        let trial = state.trials[state.currentIndex]
        eventContinuation.yield(.trial(trial, index: state.currentIndex))
    }

    private func finish() {
        state.isFinished = true
        let finishedAt = Date()
        let correctCount = state.answers.filter(\.correct).count
        let rts = state.answers.map(\.rtMs)
        let medianRt = Scoring.median(rts)
        // Aggregate flag: if the whole session sits below the human floor,
        // count one extra drift.
        if medianRt > 0, medianRt < aggregateMedianFloorMs {
            state.drifts += 1
        }
        let breakdown = Scoring.score(
            correct: correctCount,
            total: state.answers.count,
            rtMedianMs: medianRt,
            rtStddevMs: Scoring.standardDeviation(rts),
            drifts: state.drifts
        )
        let result = SessionResult(
            sessionId: state.sessionId,
            userIdentifier: state.userIdentifier,
            gameId: state.game.id,
            construct: state.game.construct,
            startedAt: state.startTime,
            finishedAt: finishedAt,
            trials: state.trials,
            answers: state.answers,
            score: breakdown.score,
            rtMedianMs: breakdown.rtMedianMs,
            rtStddevMs: breakdown.rtStddevMs,
            accuracy: breakdown.accuracy,
            drifts: breakdown.drifts
        )
        eventContinuation.yield(.finish(result))
        eventContinuation.finish()
    }

    /// Convert a `Duration` (from `ContinuousClock`) to milliseconds.
    private static func durationToMs(_ duration: Duration) -> Int {
        let comps = duration.components
        let totalSeconds = Double(comps.seconds) + Double(comps.attoseconds) / 1e18
        return Int((totalSeconds * 1000).rounded())
    }

    /// Determines whether `response` is the correct answer for `trial`.
    /// Throws `EngineError.invalidResponse` if the response shape does not
    /// match the trial's template. This is a programming bug, not a user
    /// error — it should be caught in development, not silently scored wrong.
    private func isCorrect(trial: Trial, response: TrialResponse) throws -> Bool {
        switch (trial, response) {
        case (.choice(let t), .choice(let id)):
            return t.choices.first(where: { $0.id == id })?.correct ?? false
        case (.recall(let t), .recall(let id)):
            return t.choices.first(where: { $0.id == id })?.correct ?? false
        case (.reaction(let t), .reaction(let pressed)):
            return pressed == t.shouldPress
        case (.sequence(let t), .sequence(let user)):
            guard user.count == t.answer.count else { return false }
            return zip(user, t.answer).allSatisfy { $0 == $1 }
        case (.grid(let t), .grid(let cell)):
            return cell.row == t.answer.row && cell.col == t.answer.col
        case (.numberLine(let t), .numberLine(let v)):
            let span = t.max - t.min
            return abs(v - t.target) <= span * t.tolerance
        case (.typed(let t), .typed(let user)):
            let trimmed = user.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            return (try? NSRegularExpression(pattern: t.answerPattern, options: [.caseInsensitive]))
                .map { rx in
                    let range = NSRange(trimmed.startIndex..., in: trimmed)
                    return rx.firstMatch(in: trimmed, range: range) != nil
                } ?? false
        case (.sort(let t), .sort(let idx)):
            return idx == t.answerIndex
        default:
            // Response shape doesn't match the trial template — a programming
            // bug. We throw rather than returning false so a dev catches it.
            throw EngineError.invalidResponse
        }
    }

    private func detectDrift(rtMs: Int, response: TrialResponse) -> Bool {
        if rtMs < rtImpossibleFloorMs { return true }
        // Streak walk: if any of the last `driftStreakCap` answers have the
        // same RT and response, flag the current trial. A bot or "next" key
        // spammer produces this pattern; a human rarely does.
        for prior in state.answers.reversed().prefix(driftStreakCap) {
            if prior.rtMs == rtMs && prior.response == response {
                return true
            }
        }
        return false
    }
}
